### ISSUE TITLE
defining a validator causes mypy: maximum semantic analysis iteration count reached

### ISSUE THREAD

**Issue by bluefish6** [pydantic/pydantic#5446]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

Hi, I'm getting the mypy `INTERNAL ERROR: maximum semantic analysis iteration count reached` error when defining a validator in a model via a helper function. Note that declaring the type of the validator makes mypy not crash anymore (see example). Also the issue occurs only with `pydantic.mypy` plugin.

```
mypy tests/test_validators2.py
Deferral trace:
    tests.test_validators2:13
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
    tests.test_validators2:11
tests/test_validators2.py: error: INTERNAL ERROR: maximum semantic analysis iteration count reached
Found 1 error in 1 file (errors prevented further checking)
```

### Example Code

```Python
import pydantic

def get_my_custom_validator(field_name: str) -> classmethod:
    @pydantic.validator(field_name, allow_reuse=True)
    def my_custom_validator(cls, v: int) -> int:
        return v

    return my_custom_validator

def foo():
    class MyModel(pydantic.BaseModel):
        number: int
        custom_validator = get_my_custom_validator("number")
        # custom_validator: classmethod = get_my_custom_validator("number")  # does not crash

    pass
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.10.7
            pydantic compiled: True
                 install path: /home/.../.cache/pypoetry/virtualenvs/...-aO4xbsjX-py3.10/lib/python3.10/site-packages/pydantic
               python version: 3.10.4 (main, Apr 27 2022, 18:18:04) [GCC 9.4.0]
                     platform: Linux-5.14.0-1054-oem-x86_64-with-glibc2.35
     optional deps. installed: ['dotenv', 'typing-extensions']


mypy: both 1.0.1 and 1.1.1 give the same error


pyproject.toml configuration:

[tool.mypy]
python_version = "3.10"

plugins = [
  "pydantic.mypy"
]

check_untyped_defs = true
ignore_missing_imports = true
namespace_packages = true
strict = true

disallow_any_generics = false
disallow_incomplete_defs = false
disallow_subclassing_any = false
disallow_untyped_calls = false
disallow_untyped_decorators = false
disallow_untyped_defs = false
no_implicit_optional = false
no_implicit_reexport = false
strict_equality = false
warn_return_any = false
warn_redundant_casts = true
warn_unused_ignores = true

[tool.pydantic-mypy]
init_forbid_extra = true
init_typed = true
warn_required_dynamic_aliases = true
warn_untyped_fields = false
```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/usage/schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/usage/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/usage/model_config/)
- [ ] [Field Types](https://docs.pydantic.dev/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/usage/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [X] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by dmontagu** [comment]
I believe the issue here is coming from the fact that in v1, you can have non-annotated fields, and the mypy plugin treats those as fields. I did a bit of playing with the mypy plugin and it will be a bit hard to resolve this.

Fortunately this will be easier in v2 as we don't allow un-annotated fields. In v2 you still get a spurious error saying we don't allow un-annotated fields, but at least you don't get the INTERNAL ERROR and if you ignore that error I think everything else will behave fine.

All this said, I found a tweak that seems to make all existing tests pass, though I don't fully understand what's going on. But considering all tests pass and it prevents the INTERNAL ERROR I think it's an improvement. I'll open a PR with it shortly.

**Comment by dmontagu** [comment]
This should be fixed now in the just-released v1.10.8, which should be available through PyPI shortly.
